Robot API based on sourcebots' sbot

WillB97 opened the pull request:
A lightweight Python 3 API based on 

- Contains a serial wrapper that handles reconnecting and command retries.
- Contains and updated `Marker` API with a clearer set of fields.

RealOrangeOne commented on sr/robot/game_specific.py:
```diff
@@ -3,8 +3,7 @@
 
-GAME_LENGTH = 150  # seconds
-
+# TODO: Update this for 2023
```
```suggestion
# TODO: Update this for 2024
```

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

RealOrangeOne commented on sr/robot/utils.py:
```diff
@@ -199,3 +201,27 @@ def get_USB_identity(port: ListPortInfo) -> BoardIdentity:
         return BoardIdentity()
+
+
+def ensure_atexit_on_term() -> None:
```
This implementation is identical to https://github.com/sourcebots/sbot/pull/334, just copied. I'd suggest merging the open PR upstream and rebasing the PR. This keeps the commit history cleaner and the PR more accurate.

RealOrangeOne commented on sr/robot/robot.py:
```diff
@@ -239,32 +259,71 @@ def camera(self) -> AprilCamera:
-    def metadata(self) -> Metadata:
+    def time(self) -> float:
+        """
+        Get the number of seconds since the Unix Epoch.
```
Should we put a warning here about time on the robot not being correct, or leave that as a deployment detail?

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

WillB97 committed:
Updated Arduino API to follow the sr-robot3 API and be compatible with the SR API

Add support for custom arduino commands

WillB97 committed:
Update brain port

Downgrade pydantic to avoid collision with astoria

WillB97 committed:
Update vision API

WillB97 committed:
Add class for exposing generic serial ports
Use only serial number for raw ports

WillB97 committed:
Update MQTT integrations
Mutex setting MQTT subscriptions

WillB97 committed:
Update metadata integration

WillB97 committed:
Update kch
Support offloading flashing start LED
Move setting KCH colour into the LED class

WillB97 committed:
Support ignoring Arduinos by serial number

WillB97 committed:
Add remaining initial updates and todos

WillB97 committed:
Fix tests

WillB97 committed:
Change discovery to cope wiith errors thrown while querying identity

WillB97 committed:
Remove metadata property from robot object

WillB97 committed:
Make sure atexit runs on sigterm

WillB97 committed:
Complete todos

WillB97 committed:
Fixes from testing

WillB97 committed:
Also handle OSError in serial wrapper

WillB97 committed:
Add unit tests for the marker API (#3)

* Add CI for marker API
* Tighten bounds
* Bump april_vision

RealOrangeOne commented on sr/robot/arduino.py:
```diff
@@ -92,13 +92,41 @@ def __init__(
         self._serial.set_identity(self._identity)
 
+    @classmethod
+    def _get_valid_board(
```
Suggestion: There seems to be a lot of duplication around board classes in `_get_valid_board`. Perhaps we should move it to a mixin, or to the base class itself?

RealOrangeOne replied:
Also, the method name isn't super descriptive - perhaps a doc string?

WillB97 replied:
664b3ef70b6f6392575db650c6005aa3d7ae5cca

RealOrangeOne commented on sr/robot/arduino.py:
```diff
@@ -235,21 +248,21 @@ def __init__(
         Get the mode of the pin.
 
-        This is fetched from the board.
+        This returns the cached value since thew board does not report this.
```
```suggestion
        This returns the cached value since the board does not report this.
```

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

RealOrangeOne commented on sr/robot/arduino.py:
```diff
@@ -282,12 +304,11 @@ def digital_value(self) -> bool:
-        response = self._serial.query(f'PIN:{self._index}:DIGITAL:GET?')
-        return (response == '1')
+        response = self._serial.query(self._build_command('r'), endl='')
+        return (response == 'h')
```
```suggestion
        return response == 'h'
```

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

RealOrangeOne commented on sr/robot/astoria.py:
```diff
@@ -0,0 +1,294 @@
+        return pressed
+
+
+def init_mqtt(config: AstoriaConfig, client_name: str = 'sr-robot3') -> 'MQTTClient':
```
```suggestion
def init_mqtt(config: AstoriaConfig, client_name: str = 'sr-robot') -> 'MQTTClient':
```

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

RealOrangeOne commented on sr/robot/astoria.py:
```diff
@@ -0,0 +1,294 @@
+        self._metadata_lock = Lock()
+        self._metadata = Metadata()
+        self._mount_path = Path("/dev/null")
+        self._mount_path_lock = Lock()
```
Question: Given assignments are atomic, is there a need for a lock? Looks like we're just adding extra latency when we don't need it.

WillB97 replied:
https://github.com/google/styleguide/blob/91d6e367e384b0d8aaaf7ce95029514fcdf38651/pyguide.md#218-threading

> While Python's built-in data types such as dictionaries appear to have atomic operations, there are corner cases where they aren't atomic (e.g. if `__hash__` or `__eq__` are implemented as Python methods) and their atomicity should not be relied upon. Neither should you rely on atomic variable assignment (since this in turn depends on dictionaries).

prophile commented on sr/robot/astoria.py:
```diff
@@ -0,0 +1,294 @@
+    sender_name: str
+    priority: int = 0
+
+    def __gt__(self, other: "StartButtonBroadcastEvent") -> bool:
+        return self.priority > other.priority
```
Usually better to complete the set if we're implementing comparisons, `@functools.total_ordering` on the class perhaps?

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

prophile commented on sr/robot/astoria.py:
```diff
@@ -0,0 +1,294 @@
+        self._metadata_lock = Lock()
+        self._metadata = Metadata()
+        self._mount_path = Path("/dev/null")
+        self._mount_path_lock = Lock()
```
Could we use a single lock? Less granular but I don't think we'll have contention between them anyway and removes one potential deadlock point.

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

prophile commented on sr/robot/astoria.py:
```diff
@@ -0,0 +1,294 @@
+    def fetch_current_metadata(self) -> Metadata:
+        """Fetch the current metadata."""
+        with self._metadata_lock:
+            return self._metadata
```
It does make me a bit uncomfortable that we're doing thread safe protection but we're returning this mutable thing. How would you feel about a `deepcopy` here?

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

prophile commented on sr/robot/robot.py:
```diff
@@ -149,6 +159,15 @@ def _log_connected_boards(self) -> None:
+    @property
+    def kch(self) -> KCH:
+        """
+        Access the Raspberry Pi hat, includeing user-accessible LEDs.
```
typo: including

WillB97 replied:
28f4c9008c93f1664160a7fb1843a9786f41b83b

prophile commented on sr/robot/robot.py:
```diff
@@ -240,32 +259,71 @@ def camera(self) -> AprilCamera:
         """
-        Fetch the robot's current metadata.
+        if self._metadata is None:
+            raise MetadataNotReadyError()
```
pepega

(but I know why this is here)

prophile commented on sr/robot/servo_board.py:
```diff
@@ -87,6 +87,31 @@ def __init__(
+        cls,
+        serial_port: str,
+        initial_identity: BoardIdentity | None = None,
+    ) -> None | ServoBoard:
```
Needs to be an explicit `Union` for the lowest Pythons we support 😢

WillB97 replied:
84f98774c3c2719ea051507c2b0dc0ba67315f1d

WillB97 committed:
Python 3.8 typehints

WillB97 committed:
Add helper function to print information of attached serial ports

WillB97 committed:
Support adding extra calibration files to the package

WillB97 committed:
Add blocking argument to beep

WillB97 committed:
Update marker sizes

RealOrangeOne commented on sr/robot/arduino.py:
```diff
@@ -69,7 +70,7 @@ def get_board_type() -> str:
         self,
         serial_port: str,
-        initial_identity: BoardIdentity | None = None,
+        initial_identity: Union[BoardIdentity, None] = None,
```
Nit: `Optional[T]` is clearer than `Union[T, None]`

WillB97 replied:
bfbb2e36e3636eb9bf679f5664b74761d216df92

RealOrangeOne commented on sr/robot/power_board.py:
```diff
@@ -475,7 +476,7 @@ def __init__(self, serial: SerialWrapper):
 
     @log_to_debug
-    def buzz(self, frequency: float, duration: float) -> None:
+    def buzz(self, frequency: float, duration: float, *, blocking: bool = False) -> None:
```
TODO: Document this new argument

WillB97 replied:
bfbb2e36e3636eb9bf679f5664b74761d216df92

RealOrangeOne commented on sr/robot/camera.py:
```diff
@@ -55,7 +58,7 @@ def _discover(cls) -> Dict[str, 'AprilCamera']:
             (serial := f"{camera_data.name} - {camera_data.index}"):
             cls(camera_data.index, camera_data=camera_data, serial_num=serial)
-            for camera_data in find_cameras(calibrations)
+            for camera_data in find_cameras(robot_calibrations)
```
Suggestion: Should we also include the user code directory here so competitors can easily add their own calibrations?

WillB97 replied:
`calibrations` already includes this, I was inserting the sr-robot calibrations below the april_vision calibrations though so I've corrected that.
bfbb2e36e3636eb9bf679f5664b74761d216df92

WillB97 committed:
Review fixes

8BitJosh commented on sr/robot/power_board.py:
```diff
@@ -477,6 +485,7 @@ def buzz(self, frequency: float, duration: float) -> None:
+        :param blocking: Whether to block until the tone has finished playing,
         """
         frequency_int = int(float_bounds_check(
             frequency, 8, 10_000, "Frequency is a float in Hz between 0 and 10000"))
```
This error message is incorrect, it is between 8 and 10,000

RealOrangeOne replied:
```suggestion
            frequency, 8, 10_000, "Frequency must be between 8 and 10000Hz"))
```

RealOrangeOne commented on sr/robot/arduino.py:
```diff
@@ -161,15 +176,18 @@ def identify(self) -> BoardIdentity:
         return BoardIdentity(
-            manufacturer=response_fields[0],
-            board_type=response_fields[1],
+            manufacturer=(
+                "Student Robotics"
+                if response_fields[0].startswith('SR')
+                else "Arduino"),
+            board_type=response_fields[0],
```
Nit: The board is always manufactured by "Arduino". We should have another way of detecting custom firmware

WillB97 replied:
We don't need to detect custom firmware because from our perspective it makes no difference. It's just for indication.

RealOrangeOne commented on sr/robot/arduino.py:
```diff
@@ -267,11 +281,20 @@ def mode(self, value: GPIOPinMode) -> None:
-        self._serial.write(f'PIN:{self._index}:MODE:SET:{value.value}')
 
-    @property
+        mode_char_map = {
```
Nit: This would make a good constant

RealOrangeOne commented on sr/robot/astoria.py:
```diff
@@ -0,0 +1,295 @@
+else:
+    import tomli as tomllib
+
+astoria_version = '0.10.0'
```
Nit: Constant should be upper-case

RealOrangeOne commented on sr/robot/astoria.py:
```diff
@@ -0,0 +1,295 @@
+    mqtt: MQTTBrokerInfo
+
+    @classmethod
+    def _get_config_path(cls, config_str: Optional[str] = None) -> Path:
```
Nit: `config_str` isn't a very descriptive name. It implies it _is_ the config, rather than the path to it. Perhaps `config_path`?

WillB97 replied:
5396f226c2f39632d0cbe898334b1b033ec8efa4

RealOrangeOne commented on sr/robot/kch.py:
```diff
@@ -0,0 +1,188 @@
+        # We are not running cleanup so the LED state persists after the code completes,
+        # this will cause a warning with `GPIO.setup()` which we can suppress
+        if HAS_HAT:
+            # atexit.register(GPIO.cleanup)
```
Nit: Commented out code

WillB97 replied:
The code is an example of what you would normally do and the comment above explains why we're not doing that.
I've rearranged it in 5396f226c2f39632d0cbe898334b1b033ec8efa4 to be clearer.

RealOrangeOne commented on sr/robot/timeout.py:
```diff
@@ -56,4 +56,4 @@ def kill_after_delay(timeout_seconds: int) -> None:
         signal.signal(signal.SIGALRM, timeout_handler)
         signal.alarm(timeout_seconds)
-    logger.debug(f"Kill Signal Timeout set: {timeout_seconds}s")
+    logger.info(f"Kill Signal Timeout set: {timeout_seconds}s")
```
Nit: The language in this message erks me

```suggestion
    logger.info(f"Timeout set: {timeout_seconds}s")
```

8BitJosh commented on sr/robot/utils.py:
```diff
@@ -216,14 +216,31 @@ def ensure_atexit_on_term() -> None:
+    for serial_port in comports():
+        if serial_port.vid and serial_port.pid:
+            # Only print USB serial ports
+            print(
```
Should we add a line before this print to make it more clear what is being printed
e.g. 
`print("VID:PID - Manufacturer - Product - Serial Number")`

8BitJosh commented on sr/robot/kch.py:
```diff
@@ -0,0 +1,188 @@
+            atexit.register(GPIO.cleanup, RobotLEDs.START)  # type: ignore[call-arg]
+
+    @property
+    def start(self) -> bool:
```
methods to do with start LED should be "private"

8BitJosh commented:
robot.camera.see

`def see(self, *, eager: bool = True, frame: Optional[NDArray] = None) -> List[Marker]:`

Unused argument "eager"
This is no longer used and should be removed

8BitJosh commented:
The raw serial module is got some issues with the design.

Generally with a read/write/query interface the behavior should be:
Write = send some data 
Read = read in some data
Query = send some data then read in some data (equivalent to write + read)

Currently our system:
- Write = send some data which we terminate with a newline then discard the response that is also terminated with a newline, crash/retry if there is no response
- Read = read some data terminated with a newline
- Query = send some data which we terminate with a newline then return the response that is also terminated with a newline

In this case the behavior of read and write is weird.
If you do a write and a read, instead of a query, the read will return nothing.
The behavior of the query =/= write + read

This system also constrains the user/competitor to a very strict serial structure with everything newline terminated and writes also expecting an ACK/NACK response.

8BitJosh commented on sr/robot/kch.py:
```diff
@@ -0,0 +1,188 @@
+                self._pwm = None
+            GPIO.output(RobotLEDs.START, GPIO.HIGH if value else GPIO.LOW)
+
+    def flash_start(self) -> None:
```
flash_start -> _flash_start

WillB97 committed:
Remove problem parts of raw serial
For the moment there is no reconnection logic

WillB97 committed:
Allow log_to_debug to not print return values of setters
